exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/ColumnType.kt

```diff
@@ -1021,29 +975,42 @@ class CustomEnumerationColumnType<T : Enum<T>>(
     val fromDb: (Any) -> T,
     /** Returns the function that converts an enumeration instance [T] to a value that will be stored to a database. */
     val toDb: (T) -> Any
-) : StringColumnType() {
+) : ColumnType<T>() {
     override fun sqlType(): String = sql ?: error("Column $name should exist in database")
 
     @Suppress("UNCHECKED_CAST")
     override fun valueFromDB(value: Any): T = if (value::class.isSubclassOf(Enum::class)) value as T else fromDb(value)
 
-    @Suppress("UNCHECKED_CAST")
-    override fun notNullValueToDB(value: Any): Any = toDb(value as T)
+    override fun notNullValueToDB(value: T): Any = toDb(value)
 
-    override fun nonNullValueToString(value: Any): String = super.nonNullValueToString(notNullValueToDB(value))
+    override fun nonNullValueToString(value: T): String = buildString {
+        append('\'')
+        append(escape(notNullValueToDB(value).toString()))
+        append('\'')
+    }
+
+    private fun escape(value: String): String = value.map { charactersToEscape[it] ?: it }.joinToString("")
+
+    companion object {
+        private val charactersToEscape = mapOf(
+            '\'' to "\'\'",
+            '\r' to "\\r",
+            '\n' to "\\n"
+        )
+    }
 }
 
 // Array columns
 
 /**
  * Array column for storing a collection of elements.
  */
-class ArrayColumnType(
+class ArrayColumnType<E>(
     /** Returns the base column type of this array column's individual elements. */
-    val delegate: ColumnType,
+    val delegate: ColumnType<E & Any>,
     /** Returns the maximum amount of allowed elements in this array column. */
     val maximumCardinality: Int? = null
-) : ColumnType() {
+) : ColumnType<List<E?>>() {
```

Comment:
Breaking change

Reply:
Is it necessary that the returned type be a list of nullables?

Before, the contents of the input list could be restricted based on the type provided to `array()` and this would be enforced by the insert lambda. Now nulls can always be inserted.
```kt
val numbers: Column<List<Int>> = array<Int>("numbers")
val nullNumbers: Column<List<Int?>> = array<Int?>("null_numbers")

val input = listOf(1, null, 3) // this could only be inserted in the second column before
```